app/models/grda_warehouse/hud/client.rb

```diff
@@ -2027,6 +2027,28 @@ def total_adjusted_days enrollments
     def total_months enrollments
       enrollments.map{|e| e[:months_served]}.flatten(1).uniq.size
     end
+    
+    def affiliated_projects_str_for_enrollment enrollment
+      affiliated_project_names = GrdaWarehouse::Hud::Project.find_by(ProjectID: enrollment[:ProjectID]).affiliated_projects.pluck(:ProjectName)
```

Comment:
This is doing 2 queries (and not obeying the multi-key relationship between enrollment and project).  It could be re-written as something like `GrdaWarehouse::Hud::Affiliation.joins(:project).where(ProjectID: enrollment[:ProjectID], data_source_id: enrollment[:data_source_id]).pluck(p_t[:ProjectName].to_sql)`  (Same for below but with a different join.

Reply:
Also, this should probably take an array of enrollments so you can get the entire graph  back in one query, then index on enrollment[:id] (or something similar).  As adjusted above, this would be one query per-enrollment per affiliation type, which could be hundreds of queries for some clients. 

Better to pass in an array of hashes of the following structure `[{enrollment_id: 1, ProjectID: 2, data_source_id: 3},]` and then do  the one query for all and group by enrollment_id.